src/main/java/com/commercetools/sync/customers/helpers/CustomerBatchValidator.java

```diff
@@ -0,0 +1,223 @@
+    public ImmutablePair<Set<CustomerDraft>, ReferencedKeys> validateAndCollectReferencedKeys(
+        @Nonnull final List<CustomerDraft> customerDrafts) {
+
+        final ReferencedKeys referencedKeys = new ReferencedKeys();
+        final Set<CustomerDraft> validDrafts = customerDrafts
+            .stream()
+            .filter(this::isValidCustomerDraft)
+            .peek(customerDraft -> collectReferencedKeys(referencedKeys, customerDraft))
+            .collect(toSet());
+
+        return ImmutablePair.of(validDrafts, referencedKeys);
+    }
+
+    private boolean isValidCustomerDraft(@Nullable final CustomerDraft customerDraft) {
+
+        if (customerDraft == null) {
+            handleError(CUSTOMER_DRAFT_IS_NULL);
+        } else if (isBlank(customerDraft.getKey())) {
+            handleError(format(CUSTOMER_DRAFT_KEY_NOT_SET, customerDraft.getEmail()));
+        } else if (isBlank(customerDraft.getEmail())) {
+            handleError(format(CUSTOMER_DRAFT_EMAIL_NOT_SET, customerDraft.getKey()));
+        } else if (hasValidAddresses(customerDraft)) {
+            return hasValidBillingAndShippingAddresses(customerDraft);
+        }
+        return false;
+    }
+
+    private boolean hasValidAddresses(@Nonnull final CustomerDraft customerDraft) {
+
+        final List<Address> addressList = customerDraft.getAddresses();
+        if (addressList == null || addressList.isEmpty()) {
+            return true;
+        }
+
+        final List<Integer> indexesOfNullAddresses = getIndexes(addressList, Objects::isNull);
+        if (!indexesOfNullAddresses.isEmpty()) {
+            handleError(format(ADDRESSES_ARE_NULL, customerDraft.getKey(), indexesOfNullAddresses));
+            return false;
+        }
+
+        final List<Integer> indexesOfBlankAddressKeys = getIndexes(addressList, address -> isBlank(address.getKey()));
+        if (!indexesOfBlankAddressKeys.isEmpty()) {
+            handleError(format(ADDRESSES_THAT_KEYS_NOT_SET, customerDraft.getKey(), indexesOfBlankAddressKeys));
+            return false;
+        }
+
+        final Predicate<Address> searchDuplicateKeys = (Address address) ->
+            addressList.stream().filter(a -> Objects.equals(a.getKey(), address.getKey())).count() > 1;
+
+        final List<Integer> indexesOfDuplicateAddressKeys = getIndexes(addressList, searchDuplicateKeys);
+        if (!indexesOfDuplicateAddressKeys.isEmpty()) {
+            handleError(format(ADDRESSES_THAT_KEYS_NOT_UNIQUE, customerDraft.getKey(), indexesOfDuplicateAddressKeys));
+            return false;
+        }
+
+        return true;
+    }
+
+    @Nonnull
+    private List<Integer> getIndexes(@Nonnull final List<Address> list, @Nonnull final Predicate<Address> predicate) {
+
+        final List<Integer> indexes = new ArrayList<>();
+        for (int i = 0; i < list.size(); i++) {
+            if (predicate.test(list.get(i))) {
+                indexes.add(i);
+            }
+        }
+        return indexes;
```

Comment:
```suggestion
        return list.stream().filter(predicate).collect(Collectors.toList());
```

Reply:
@leungkinghin do you think one-liner will do the same? because method returns integer values (indexes), not addresses?